### PR TITLE
feat: Pattern modules for policy exemptions

### LINKED ISSUE #1735: [Module Proposal]: `avm/ptn/authorization/policy-exemption`
### Check for previous/existing GitHub issues/module proposals

- [x] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [x] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

Pattern Module

### Module Name

avm/ptn/authorization/policy-exemption

### Module Details

Create pattern module to allow the deployment of policy exemptions at management group, subscription level, and resource group scopes.

### Do you want to be the owner of this module?

Yes

### Module Owner's GitHub Username (handle)

oZakari

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by oZakari**
## Description

Adds the pattern modules for Policy Exemptions

Closes https://github.com/Azure/Azure-Verified-Modules/issues/1735
## Pipeline Reference

| Pipeline |
| -------- |
|  [avm.ptn.authorization.policy-exemption](https://github.com/oZakari/bicep-registry-modules/actions/workflows/avm.ptn.authorization.policy-exemption.yml)   |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings

### CODE REVIEW

**Comment by oZakari** [comment]
Hi team, I assume the telemetry errors and the module identifier errors are related to the directory structure for this pattern module (similar to the ALZ Policy Assignments Pattern module) and this [file](https://github.com/Azure/Azure-Verified-Modules/blob/main/docs/static/module-indexes/BicepPatternModules.csv)? Anything I need to do on my side?

**Comment by oZakari** [comment]
> Hi team, I assume the telemetry errors and the module identifier errors are related to the directory structure for this pattern module (similar to the ALZ Policy Assignments Pattern module) and this [file](https://github.com/Azure/Azure-Verified-Modules/blob/main/docs/static/module-indexes/BicepPatternModules.csv)? Anything I need to do on my side?

Any update on this one?

**Comment by AlexanderSehr** [comment]
Hey @oZakari,
sorry has been a busy week. Will try to squeeze the review in

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:9]: Either the team is not created or not yet approved in the hierachy ([ref](https://azure.github.io/Azure-Verified-Modules/specs/shared/#id-snfr20---category-contributionsupport---github-teams-only))

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/policy-exemption/main.bicep:2]: On a different note, has there been an AVM module proposal for this? I have not found the module in the MAR file which is a prerequisite to successfully publish.

cc: @matebarabas

File: avm/ptn/authorization/policy-exemption/tests/e2e/mg.defaults/main.test.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/policy-exemption/tests/e2e/mg.defaults/main.test.bicep:10]: This is ideally directly implemented in line 29 instead of introducing a new parameter/variable which is only used in a single location.

File: avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep
Comment (AlexanderSehr) [inline avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep:16]: Same comment as in min file

Comment (AlexanderSehr) [reply avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep:16]: I'll not comment this in every file now, so either please respond in one of the 2 comments, or if you agree with the suggestion, please implement it for the other test files too :)

File: .github/CODEOWNERS
Comment (oZakari) [reply .github/CODEOWNERS:9]: Got it, must be that it hasn't been approved yet then as it exists [here](https://github.com/orgs/Azure/teams/avm-ptn-authorization-policy-exemption-module-owners-bicep).

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/main.bicep:2]: Mate had me create 3 separate ones: 

-https://github.com/Azure/Azure-Verified-Modules/issues/1405?notification_referrer_id=NT_kwDOAddDN7QxMjQzNDk1NzkwMzozMDg4NDY2Mw#issuecomment-2422706165
-https://github.com/Azure/Azure-Verified-Modules/issues/1420?notification_referrer_id=NT_kwDOAddDN7QxMjQ5MjIxNzYwMjozMDg4NDY2Mw#issuecomment-2379474840
-https://github.com/Azure/Azure-Verified-Modules/issues/1421?notification_referrer_id=NT_kwDOAddDN7QxMjQ5MjIyNjAyMTozMDg4NDY2Mw#issuecomment-2379473633

File: avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep
Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep:16]: Think I'm going to keep as is to avoid breaking this linter rule: alt text as I think it'll be confusing for people who are using custom policies. Open to discussing though.

Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep:16]: Feel free to reopen if you want to discuss more.

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/main.bicep:1]: Is it okay to go with this approach for now and then refactor in the future? Reason being that I would like to use this for the refactor of ALZ-Bicep which is currently in progress.

**Comment by oZakari** [comment]
> Hey @oZakari, sorry has been a busy week. Will try to squeeze the review in

No worries at all, @AlexanderSehr! Thank you for taking the time to review it—I just wanted to make sure it was on your radar. Really appreciate it! 😊

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/policy-exemption/main.bicep:1]: It is, yes. Just consider it a word of warnung, I guess 😄 

The transition would then look like this
1. This module is published
2. When the scoped-modules CI feature is live, we can get started on splitting the module into 3 submodules (if desired)
3. Once published, we'd unlist the original module (note, the module will still continue to work and not removed. It would just not be showing in the 'available modules' dropdown)

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/policy-exemption/main.bicep:2]: @matebarabas we need to have a word 😄 Will ping you internally. @oZakari no worries, your module can stay as is, but something in the back needs updating.

File: .github/CODEOWNERS
Comment (AlexanderSehr) [reply .github/CODEOWNERS:9]: Hey @oZakari, please don't resolve a comment unless it's a code suggestion 😉 Makes re-reviews a bit harder.
In this case this must remain a blocking comment until the group was approved. It could lead to issues if we merge the module before.

**Comment by oZakari** [comment]
Closing this as virtual networks can now be updated without modifying associated subnets. This is capable as of API version 2023-09-01 and ALZ-Bicep is currently set to 2024-01-01. Please see this [blog ](https://techcommunity.microsoft.com/blog/azurenetworkingblog/azure-virtual-network-now-supports-updates-without-subnet-property/4067952)for additional info. Note, the virtual network module we will be transitioning to for AVM also takes this change into account.

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:1]: Please note, the `main.json` templates **must** be re-build with the latest Bicep version `0.31.34` as this version (running on the agents) renders a different `main.json` file then previous versions. 

The steps to execute would be
- Upgrade your Bicep CLI version to `0.31.34` (e.g. via `choco upgrade bicep` if choco is used)
- Re-run the `Set-AVMModule` script on your module. E.g. via `Set-AVMModule -ModuleFolderPath '.\avm\res\automation\automation-account\' -SkipFileAndFolderSetup -SkipReadMe -Recurse`
- Push the changes
- Re-run the tests to ensure everything works. 

If the static tests continue to fail for you, please note that there is currently presumably another new bug which affects some of our modules ([ref](https://github.com/Azure/bicep/issues/15517)).

Comment (oZakari) [reply .github/CODEOWNERS:1]: Done

File: .github/CODEOWNERS
Comment (oZakari) [reply .github/CODEOWNERS:9]: Bumping as tests are still failing for group approval. @matebarabas

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/policy-exemption/main.bicep:2]: Hey @matebarabas, has there bin an update already? :)

**Comment by jtracey93** [comment]
Hey @oZakari ,

Firstly, thanks for your work on this PR!

We have made some changes to the AVM CI, detailed below, which means we need you to update your fork to pull in these latest changes and re-run your tests to show they still are passing prior to approving and merging this PR, as we don't and it fails once merged the publishing of your module will fail and will be blocked going forward until the test pass again via additional PRs.

### Changes to CI That Have Been Made That You Need To Take Action On

- https://github.com/Azure/bicep-registry-modules/pull/3745
    - Adds WAF Security PSRule Checks using this custom baseline of rules ([`/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml`](https://github.com/Azure/bicep-registry-modules/blob/main/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/cb-waf-security.Rule.yaml))
    - **Action Required By You:**
        - Ingest/pull-in these changes into your PR branch and re-run your tests to ensure they pass
            - If any issues with the rules and you think it should apply, please ping/reach out to the core team

Any questions reach out to the AVM Core Team by tagging us in your PR here or internally via Teams

Thanks

Jack (AVM Core Team)

File: .github/CODEOWNERS
Comment (AlexanderSehr) [inline .github/CODEOWNERS:1]: Please note @oZakari, the static tests need to run and pass two in order to publish the module 😉

File: avm/ptn/authorization/policy-exemption/main.bicep
Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/main.bicep:2]: This has been resolved.

**Comment by oZakari** [comment]
Hey @AlexanderSehr, should be good for you to review now. Thanks!

File: avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep
Comment (AlexanderSehr) [reply avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep:16]: Hey @oZakari,
unfortunately I would 😄 But it's not a big one.

The warning you're avoidining by moving the reference to a parameter is not exactly 'addessed' doing so 😄 What Bicep wants you to do is to add a reference 

```bicep
resource policyDefinition 'Microsoft.Authorization/policyDefinitions@2025-01-01' existing = {
  name: '06a78e20-9358-41c9-923c-fb736d382a4d'
  scope: tenant()
}

# Example ref
output test string = policyDefinition.id
```

to the file and reference its ID instead.

The above example would return `/providers/Microsoft.Authorization/policyDefinitions/06a78e20-9358-41c9-923c-fb736d382a4d`.

Personally I'm not the biggest fan of that Linter rule (or at least that it applies for every ID reference), but it also has its merits.

File: .github/CODEOWNERS
Comment (oZakari) [reply .github/CODEOWNERS]: I believe this is all resolved, but will let you close @AlexanderSehr

File: avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep
Comment (oZakari) [reply avm/ptn/authorization/policy-exemption/tests/e2e/mg.max/main.test.bicep]: Ah that makes more sense. Appreciate the explanation. I have resolved the issue. Let me know if you need anything else, as hoping to get this closed out this week if possible ☺️

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**
